scripts/feature_toggle_report_generator.py

```diff
@@ -0,0 +1,237 @@
+            if toggle_type in self.toggle_states.keys():
+                self.toggle_states[toggle_type].append(toggle)
+            else:
+                self.toggle_states[toggle_type] = [toggle]
+
+    def link_toggles_to_annotations(self):
+        """
+        Read the code annotation file specified at `annotation_report_path`,
+        linking annotated feature toggles to the feature toggle state
+        entries in this Idas toggle_state dictionary.
+        """
+        if not self.annotation_report_path:
+            return
+        with io.open(self.annotation_report_path, 'r', encoding='utf-8') as annotation_file:
+            annotation_contents = yaml.safe_load(annotation_file.read())
+
+    def toggles_by_type(self, toggle_type):
+        """
+        Return a list of feature toggles for this IDA, given a type
+        of toggle (i.e. waffle_switch). Although you can just index
+        the dictionary directly, this method will handle the case when a
+        key is not present, as it will be called from Jinja templates
+        """
+        try:
+            return self.toggle_states[toggle_type]
+        except KeyError:
+            return []
+
+
+class ToggleState(object):
+    """
+    Represents an individual feature toggle within an Ida, including all
+    of it's state, pulled from the Ida's database.
+    """
+
+    def __init__(self, name, toggle_type, data):
+        self.name = name
+        self.toggle_type = toggle_type
+        self.data = data
+        self.annotation_link = None
+
+    @property
+    def state(self):
+        """
+        Return the overall state of the toggle. In other words, is it on or off
+        """
+
+        def bool_for_null_numbers(n):
+            if n == 'null':
+                return False
+            else:
+                return int(n) > 0
+
+        if self.toggle_type == 'waffle.switch':
+            return self.data['active']
+        elif self.toggle_type == 'waffle.flag':
+            return (
+                self.data['everyone'] or
+                bool_for_null_numbers(self.data['percent']) or
+                self.data['testing'] or
+                self.data['superusers'] or
+                self.data['staff'] or
+                self.data['authenticated'] or
+                bool(self.data['languages']) or
+                self.data['rollout']
+            )
+
+    @property
+    def data_for_template(self):
+        """
+        Return a dictionary of this Toggle's data for the report, formatted for
+        readability
+        """
+        null_or_number = lambda n: n if isinstance(n, int) else 0
+        template_data = {}
+        date_template = '%Y-%m-%d %H:%M'
+        if self.toggle_type == 'waffle.switch':
+            template_data['note'] = self.data['note']
+            template_data['creation_date'] = datetime.datetime.strftime(
+                self.data['created'], date_template
+            )
```

Comment:
Should all the date formatting be done by a helper function?  Might simplify the code a little.